### DISCUSSION TITLE
[rocalution] Illegal instruction detected with gfx1010

### DISCUSSION THREAD

**Discussion by IshiRenga** [rocm-arch/rocm-arch#849] [Support]
When compiling `rocalution` with `AMDGPU_TARGETS=gfx1010`, I am receiving this error which seems to be the same as this issue [here](https://github.com/rocm-arch/rocm-arch/issues/791):
```
1 warning generated when compiling for gfx1010.
error: Illegal instruction detected: Invalid dpp_ctrl value: broadcasts are not supported on GFX10+
renamable $vgpr4 = V_MOV_B32_dpp undef $vgpr4(tied-def 0), $vgpr3, 322, 10, 15, 0, implicit $exec
```

It's occurring when I'm trying to get `rocm-hip-sdk` and `rocm-opencl-sdk` from the AUR.

**Comment by reffu42** [comment]
I found a somewhat-hacky workaround for this, someone could probably make it a patch, though I just manually modified the file and built with `makepkg -e`

Edit `src/base/hip/hip_utils.hpp` line 209: Replace `#ifndef __gfx1030__` with `#ifndef __gfx1010__`

It's hacky, but it allowed the compile to work for me

Probably the ideal fix would be a patch that updates that to 1030 or 1010

**Reply by tpkessler** [reply]
I'm hesitating to add this to the repo as it has the potential to break the package. The best option would be to open an issue at https://github.com/ROCmSoftwarePlatform/rocALUTION/issues and ask for assistance. Hopefully, they already know how to fix it or may at least explain why there won't be support for gfx1010 in the future.

**Comment by JP-Ellis** [comment]
Can you check whether this results in a usable `rocaluation`? It is possible that the compilation works, but then assumes the GPU has some instructions which actually aren't there.

**Comment by xuhuisheng** [comment]
pytorch and tensorflow-rocm didn't need rocalution right now. if you don't need rocalution exactly, it unecessery to building rocalution.

RDNA lacks dpermute related instructions. because of wavefront=32, the fix for gfx1030 may suiteable for gfx1010 and other RDNA cards.

**Comment by IshiRenga** [comment]
Thank you @reffu42 , that does allow it to build and everything else got installed properly. 
@JP-Ellis As for if it built a working version, going off of the test provided by the [docs](https://rocalution.readthedocs.io/en/master/install.html#simple-test), it says the backend was CPU. I don't know what the expected output should be, but I would think it should be GPU instead, right?
```
> ./cg gr_30_30.mtx 
Number of CPU cores: 12
Host thread affinity policy - thread mapping on every core
Number of HIP devices in the system: 1
rocALUTION ver 2.0.2-ffea2fa
rocALUTION platform is initialized
Accelerator backend: HIP
OpenMP threads: 12
rocBLAS ver 2.44.0.fac8f41
rocSPARSE ver 2.2.0-1d86614
Selected HIP device: 0
------------------------------------------------
Device number: 0
Device name: AMD Radeon RX 5700 XT
totalGlobalMem: 8176 MByte
clockRate: 2100000
compute capability: 10.1
------------------------------------------------
MPI is not initialized
ReadFileMTX: filename=gr_30_30.mtx; reading...
ReadFileMTX: filename=gr_30_30.mtx; done
LocalMatrix name=gr_30_30.mtx; rows=900; cols=900; nnz=7744; prec=64bit; format=CSR; host backend={CPU(OpenMP)}; accelerator backend={HIP}; current=HIP
PCG solver starts, with preconditioner:
Jacobi preconditioner
IterationControl criteria: abs tol=1e-15; rel tol=1e-06; div tol=1e+08; max iter=1000000
IterationControl initial residual = 33.2866
IterationControl RELATIVE criteria has been reached: res norm=2.03206e-05; rel val=6.10474e-07; iter=36
PCG ends
Solver execution:0.015984 sec
||e - x||_2 = 8.01194e-06
```
@xuhuisheng Yes I am trying to get pytorch to work, but it's still giving me issues after installing the 2 packages in the OP and pipping from the pytorch website.

```
>>> import torch
>>> torch.cuda.is_available()
"hipErrorNoBinaryForGpu: Unable to find code object for all current devices!"
Aborted (core dumped)
```
What fix are you talking about for gfx1030 and is it something I can try on my end? I'm not very knowledgeable about this repo, so I apologize in advance.

**Comment by reffu42** [comment]
It does work, I've actually managed to get stable-diffusion working and have been testing that a bit this afternoon

**Comment by reffu42** [comment]
From what I can tell, rocSPARSE at one point had a similar issue, that was fixed with  [this commit](https://github.com/ROCmSoftwarePlatform/rocSPARSE/commit/f8934f91f779c291a5cf1157ed58fc427544fd2d). Presumably, rocalution would need something similar

**Comment by reffu42** [comment]
@IshiRenga 
At the risk of derailing this issue thread too much, to get around the hiperror issue, I had to build [pytorch](https://aur.archlinux.org/packages/python-pytorch-rocm) (and [torchvision](https://aur.archlinux.org/packages/python-torchvision-rocm)) from the AUR Package and make sure I set `export PYTORCH_ROCM_ARCH=gfx1010` before starting the install/build
